File: src/demo/demo_client.cpp
#include "ResourceManagerClient.h"

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace mu2e;

static const char* GREEN  = "\033[32m";
static const char* RED    = "\033[31m";
static const char* YELLOW = "\033[33m";
static const char* BOLD   = "\033[1m";
static const char* RESET  = "\033[0m";
static const char* DIM    = "\033[2m";

static void section(int n, const std::string& title)
{
    std::cout << '\n'
              << BOLD << std::string(60, '-') << RESET << '\n'
              << BOLD << " Step " << n << ": " << title << RESET << '\n'
              << BOLD << std::string(60, '-') << RESET << '\n';
}

static void printResource(const Resource& r)
{
    std::string ports;
    for (size_t i = 0; i < r.location.ports.size(); ++i) {
        if (i) ports += ',';
        ports += std::to_string(r.location.ports[i]);
    }
    const char* sc = (r.status == "available") ? GREEN : RED;
    std::cout << "  " << r.resource_class << ':' << r.name << ':' << r.enumerator
              << "  @  " << r.location.node << "  [" << ports << "]"
              << "  →  " << sc << r.status << RESET;
    if (!r.owner.empty())
        std::cout << DIM << "  (owned by " << YELLOW << r.owner << RESET << DIM << ")" << RESET;
    std::cout << '\n';
}

int main(int argc, char* argv[])
{
    std::string host = "localhost";
    int         port = 8080;
    if (argc > 1) host = argv[1];
    if (argc > 2) port = std::stoi(argv[2]);

    const std::string CLIENT_ID = "cpp-demo-client";

    std::cout << '\n'
              << BOLD << "Mu2e DAQ Resource Manager — C++ Demo Client" << RESET << '\n'
              << "Server:    " << host << ':' << port << '\n'
              << "Client ID: " << CLIENT_ID << '\n';

    ResourceManagerClient client(host, port);

    // ── 1. Server status ──────────────────────────────────────────────────
    section(1, "Server Status");
    try {
        auto s = client.getStatus();
        std::cout << "  Total: " << s.total
                  << "  |  Available: " << GREEN << s.available << RESET
                  << "  |  Reserved: "  << RED   << s.reserved  << RESET << '\n';
    } catch (const ResourceManagerException& e) {
        std::cerr << RED << "Cannot reach server: " << e.what() << RESET << '\n';
        return 1;
    }

    // ── 2. All resources ──────────────────────────────────────────────────
    section(2, "All Resources");
    auto all = client.listResources();
    for (const auto& r : all)
        printResource(r);

    // ── 3. Available resources ────────────────────────────────────────────
    section(3, "Available Resources");
    auto available = client.listResources("available");
    if (available.empty()) {
        std::cout << "  " << RED << "No available resources." << RESET << '\n';
        return 0;
    }
    for (const auto& r : available)
        printResource(r);

    // ── 4. Reserve first two ──────────────────────────────────────────────
    section(4, "Reserving Resources");
    std::vector<ResourceIdentifier> toReserve;
    int count = static_cast<int>(std::min(available.size(), size_t{2}));
    for (int k = 0; k < count; ++k) {
        toReserve.push_back({available[k].resource_class,
                             available[k].name,
                             available[k].enumerator});
        std::cout << "  Requesting: " << available[k].resource_class
                  << ':' << available[k].name
                  << ':' << available[k].enumerator << '\n';
    }

    auto result = client.reserve(CLIENT_ID, toReserve);
    if (result.success) {
        std::cout << '\n' << GREEN << "✓ " << result.message << RESET << '\n';
        for (const auto& r : result.resources)
            printResource(r);
    } else {
        std::cout << '\n' << RED << "✗ " << result.message << RESET << '\n';
        return 1;
    }

    // ── 5. Attempt double-reservation ─────────────────────────────────────
    section(5, "Attempting Double-Reservation (Expected Error)");
    std::cout << "  Requesting: " << toReserve[0].resource_class
              << ':' << toReserve[0].name << ':' << toReserve[0].enumerator
              << " as 'another-client'\n";

    auto conflictResult = client.reserve("another-client", {toReserve[0]});
    if (!conflictResult.success) {
        std::cout << '\n' << RED << "✓ Got expected error:\n"
                  << "  " << conflictResult.message << RESET << '\n';
        for (const auto& r : conflictResult.resources)
            printResource(r);
    } else {
        std::cout << YELLOW << "  Unexpected success — server did not detect conflict." << RESET << '\n';
    }

    // ── 6. Attempt non-existent resource ──────────────────────────────────
    section(6, "Attempting Non-Existent Resource (Expected Error)");
    std::cout << "  Requesting: BOGUS:FAKE:99\n";
    auto nonexist = client.reserve(CLIENT_ID, {{"BOGUS", "FAKE", "99"}});
    if (!nonexist.success)
        std::cout << '\n' << RED << "✓ Got expected error:\n  " << nonexist.message << RESET << '\n';

    // ── 7. Current state ──────────────────────────────────────────────────
    section(7, "Current Resource State (Mixed)");
    for (const auto& r : client.listResources())
        printResource(r);

    // ── 8. Release ───────────────────────────────────────────────────────
    section(8, "Releasing Reserved Resources");
    for (const auto& id : toReserve)
        std::cout << "  Releasing: " << id.resource_class << ':' << id.name << ':' << id.enumerator << '\n';

    std::string errMsg;
    if (client.release(CLIENT_ID, toReserve, errMsg)) {
        std::cout << '\n' << GREEN << "✓ Resources released successfully." << RESET << '\n';
    } else {
        std::cout << '\n' << RED << "✗ " << errMsg << RESET << '\n';
    }

    // ── 9. Final state ────────────────────────────────────────────────────
    section(9, "Final Resource State");
    for (const auto& r : client.listResources())
        printResource(r);

    auto finalStatus = client.getStatus();
    std::cout << "\n  " << BOLD << "Final Status:" << RESET
              << "  Total: " << finalStatus.total
              << "  |  Available: " << GREEN << finalStatus.available << RESET
              << "  |  Reserved: "  << RED   << finalStatus.reserved  << RESET << '\n';

    std::cout << '\n' << GREEN << "Demo complete." << RESET << "\n\n";
    return 0;
}

File: client/demo_client.py
#!/usr/bin/env python3
"""
Mu2e DAQ Resource Manager — Python Demo Client

Demonstrates a typical client workflow:
  1. Check server status
  2. List all resources
  3. List available resources
  4. Reserve resources
  5. Attempt conflicting reservation (expect error)
  6. Attempt reservation of non-existent resource (expect error)
  7. Show current state
  8. Release reserved resources
  9. Confirm final state
"""

import sys
import requests

BASE_URL  = "http://localhost:8080"
CLIENT_ID = "python-demo-client"

GREEN  = "\033[32m"
RED    = "\033[31m"
YELLOW = "\033[33m"
BOLD   = "\033[1m"
RESET  = "\033[0m"
DIM    = "\033[2m"


def section(n: int, title: str):
    print(f"\n{BOLD}{'─'*60}{RESET}")
    print(f"{BOLD} Step {n}: {title}{RESET}")
    print(f"{BOLD}{'─'*60}{RESET}")


def print_resource(r: dict):
    ports = ",".join(str(p) for p in r.get("location", {}).get("ports", []))
    status_colored = f"{GREEN}{r['status']}{RESET}" if r["status"] == "available" else f"{RED}{r['status']}{RESET}"
    owner = f"  {DIM}owned by {YELLOW}{r['owner']}{RESET}" if r.get("owner") else ""
    print(
        f"  {r['resource_class']}:{r['name']}:{r['enumerator']}"
        f"  @  {r['location']['node']}  [{ports}]"
        f"  →  {status_colored}{owner}"
    )


def get_resources(status=None):
    params = {"status": status} if status else {}
    r = requests.get(f"{BASE_URL}/api/resources", params=params, timeout=10)
    r.raise_for_status()
    return r.json()


def reserve(client_id: str, resources: list):
    return requests.post(
        f"{BASE_URL}/api/reserve",
        json={"client_id": client_id, "resources": resources},
        timeout=10,
    )


def release(client_id: str, resources: list):
    return requests.post(
        f"{BASE_URL}/api/release",
        json={"client_id": client_id, "resources": resources},
        timeout=10,
    )


def main():
    global BASE_URL
    if len(sys.argv) > 1:
        BASE_URL = sys.argv[1].rstrip("/")

    print(f"\n{BOLD}Mu2e DAQ Resource Manager — Python Demo Client{RESET}")
    print(f"Server:    {BASE_URL}")
    print(f"Client ID: {CLIENT_ID}")

    # ── 1. Server status ──────────────────────────────────────────────────
    section(1, "Server Status")
    try:
        s = requests.get(f"{BASE_URL}/api/status", timeout=10).json()
    except requests.ConnectionError:
        print(f"{RED}Cannot connect to {BASE_URL}. Is the server running?{RESET}")
        sys.exit(1)
    print(
        f"  Total: {s['total']}  |  "
        f"Available: {GREEN}{s['available']}{RESET}  |  "
        f"Reserved: {RED}{s['reserved']}{RESET}"
    )

    # ── 2. All resources ──────────────────────────────────────────────────
    section(2, "All Resources")
    all_resources = get_resources()
    for r in all_resources:
        print_resource(r)

    # ── 3. Available resources ────────────────────────────────────────────
    section(3, "Available Resources")
    available = get_resources(status="available")
    if not available:
        print(f"  {RED}No available resources — nothing to reserve.{RESET}")
        sys.exit(0)
    for r in available:
        print_resource(r)

    # ── 4. Reserve first two available resources ──────────────────────────
    section(4, "Reserving Resources")
    to_reserve = [
        {"resource_class": r["resource_class"], "name": r["name"], "enumerator": r["enumerator"]}
        for r in available[: min(2, len(available))]
    ]
    print(f"  Requesting {len(to_reserve)} resource(s) for '{CLIENT_ID}':")
    for tr in to_reserve:
        print(f"    → {tr['resource_class']}:{tr['name']}:{tr['enumerator']}")

    resp = reserve(CLIENT_ID, to_reserve)
    if resp.status_code == 200:
        data = resp.json()
        print(f"\n  {GREEN}✓ {data['message']}{RESET}")
        for r in data["resources"]:
            print_resource(r)
    else:
        print(f"\n  {RED}✗ Unexpected failure: {resp.json()}{RESET}")
        sys.exit(1)

    # ── 5. Attempt double-reservation ─────────────────────────────────────
    section(5, "Attempting Double-Reservation (Expected Error)")
    conflict_client = "another-client"
    print(f"  Trying to reserve {to_reserve[0]['resource_class']}:{to_reserve[0]['name']}:{to_reserve[0]['enumerator']}")
    print(f"  as client '{conflict_client}'")

    resp = reserve(conflict_client, [to_reserve[0]])
    if resp.status_code == 409:
        detail = resp.json().get("detail", {})
        print(f"\n  {RED}✓ Got expected 409 error:{RESET}")
        print(f"    {detail.get('message', '?')}")
        for r in detail.get("resources", []):
            print_resource(r)
    else:
        print(f"  {YELLOW}Unexpected response: {resp.status_code}{RESET}")

    # ── 6. Attempt non-existent resource ──────────────────────────────────
    section(6, "Attempting Non-Existent Resource (Expected Error)")
    bogus = [{"resource_class": "BOGUS", "name": "FAKE", "enumerator": "99"}]
    print("  Requesting: BOGUS:FAKE:99")

    resp = reserve(CLIENT_ID, bogus)
    if resp.status_code == 409:
        msg = resp.json().get("detail", {}).get("message", "?")
        print(f"\n  {RED}✓ Got expected error:{RESET} {msg}")
    else:
        print(f"  {YELLOW}Unexpected response: {resp.status_code}{RESET}")

    # ── 7. Current state ──────────────────────────────────────────────────
    section(7, "Current Resource State (Mixed)")
    for r in get_resources():
        print_resource(r)

    # ── 8. Release resources ──────────────────────────────────────────────
    section(8, "Releasing Reserved Resources")
    for tr in to_reserve:
        print(f"  Releasing {tr['resource_class']}:{tr['name']}:{tr['enumerator']}")

    resp = release(CLIENT_ID, to_reserve)
    if resp.status_code == 200:
        print(f"\n  {GREEN}✓ {resp.json()['message']}{RESET}")
    else:
        print(f"\n  {RED}✗ Release failed: {resp.json().get('detail')}{RESET}")

    # ── 9. Final state ────────────────────────────────────────────────────
    section(9, "Final Resource State")
    for r in get_resources():
        print_resource(r)

    s = requests.get(f"{BASE_URL}/api/status", timeout=10).json()
    print(
        f"\n  {BOLD}Final Status:{RESET}  Total: {s['total']}  |  "
        f"Available: {GREEN}{s['available']}{RESET}  |  "
        f"Reserved: {RED}{s['reserved']}{RESET}"
    )
    print(f"\n{GREEN}Demo complete.{RESET}\n")


if __name__ == "__main__":
    main()
